org.jacoco.core/src/org/jacoco/core/internal/instr/Companions.java

```diff
@@ -0,0 +1,253 @@
+			usedFields = 0;
+			lastId++;
+			final byte[] definition = createCompanionClass(getName());
+			try {
+				cls = new WeakReference<Class<?>>(
+						(Class<?>) DEFINE_CLASS.invoke(classLoader, getName(),
+								definition, 0, definition.length));
+			} catch (IllegalAccessException e) {
+				throw new RuntimeException(e);
+			} catch (InvocationTargetException e) {
+				throw new RuntimeException(e);
+			}
+		}
+
+		String getName() {
+			return COMPANION_NAME + lastId;
+		}
+	}
+
+	private final Map<ClassLoader, Companion> companions = new WeakHashMap<ClassLoader, Companion>();
+
+	private final RuntimeData runtimeData;
+
+	public Companions(final RuntimeData runtimeData) {
+		this.runtimeData = runtimeData;
+	}
+
+	/**
+	 * Creates instrumented version of the given class.
+	 *
+	 * @param classLoader
+	 *            defining loader
+	 * @param className
+	 *            name of the class
+	 * @param buffer
+	 *            definition of the class
+	 * @return instrumented definition
+	 */
+	public byte[] instrument(final ClassLoader classLoader,
+			final String className, final byte[] buffer) {
+		final boolean java9 = Java9Support.isPatchRequired(buffer);
+		final ClassReader reader = new ClassReader(
+				java9 ? Java9Support.downgrade(buffer) : buffer);
+
+		final int probeCount = ProbeArrayStrategyFactory.getProbeCounter(reader)
+				.getCount();
+		if (probeCount == 0) {
+			return buffer;
+		}
+		final long classId = CRC64.checksum(reader.b);
+		final IProbeArrayStrategy strategy = createStrategyFor(classLoader,
+				classId, className, probeCount);
+		final ClassWriter writer = new ClassWriter(reader, 0);
+		final ClassVisitor visitor = new ClassProbesAdapter(
+				new ClassInstrumenter(strategy, writer), true);
+		reader.accept(visitor, ClassReader.EXPAND_FRAMES);
+
+		final byte[] result = writer.toByteArray();
+		if (java9) {
+			Java9Support.upgrade(result);
+		}
+		return result;
+	}
+
+	private IProbeArrayStrategy createStrategyFor(final ClassLoader classLoader,
+			final long classId, final String className, final int probeCount) {
+		final Class<?> cls;
+		final String companionName;
+		final int fieldId;
+		synchronized (companions) {
+			Companion companion = companions.get(classLoader);
+			if (companion == null) {
+				companion = new Companion();
+				companion.defineNew(classLoader);
+				companions.put(classLoader, companion);
+			} else if (companion.usedFields == FIELDS_PER_CLASS) {
+				companion.defineNew(classLoader);
+			}
+
+			cls = companion.cls.get();
+			// if class loader is alive, then class is also alive
+			assert cls != null;
+			companionName = companion.getName();
+			fieldId = companion.usedFields;
+			companion.usedFields++;
+		}
+
+		final boolean[] probes = runtimeData
+				.getExecutionData(classId, className, probeCount).getProbes();
+		try {
+			// TODO(Godin): can cache "getMethod", if we'll use same strategy as
+			// for access of runtime data - via "equals" method
```

Comment:
Then I'd suggest removing the TODO.  If you want to keep a comment, then I would suggest mentioning there is no evidence yet that it's worthwhile and perhaps include details on how it could be implemented.